farmerbot/internal/types.go

```diff
@@ -0,0 +1,194 @@
+	SRU          uint64   `json:"sru,omitempty"`
+	CRU          uint64   `json:"cru,omitempty"`
+	MRU          uint64   `json:"mru,omitempty"`
+}
+
+// UpdateResources updates the node resources from zos resources stats
+func (n *node) updateResources(stats zos.Counters) {
+	n.resources.total.update(stats.Total)
+	n.resources.used.update(stats.Used)
+	n.resources.system.update(stats.System)
+	n.publicIPsUsed = stats.Used.IPV4U
+}
+
+// IsUnused node is an empty node
+func (n *node) isUnused() bool {
+	resources := n.resources.used.subtract(n.resources.system)
+	return resources.isEmpty() && !n.hasActiveRentContract
+}
+
+// CanClaimResources checks if a node can claim some resources
+func (n *node) canClaimResources(cap capacity, overProvisionCPU float32) bool {
+	free := n.freeCapacity(overProvisionCPU)
+	return n.resources.total.cru >= cap.cru && free.cru >= cap.cru && free.mru >= cap.mru && free.hru >= cap.hru && free.sru >= cap.sru
+}
+
+// ClaimResources claims the resources from a node
+func (n *node) claimResources(c capacity) {
+	n.resources.used.add(c)
+}
+
+// FreeCapacity calculates the free capacity of a node
+func (n *node) freeCapacity(overProvisionCPU float32) capacity {
+	total := n.resources.total
+	total.cru = uint64(math.Ceil(float64(total.cru) * float64(overProvisionCPU)))
+	return total.subtract(n.resources.used)
+}
+
+// ConsumableResources for node resources
+type consumableResources struct {
+	total  capacity
+	used   capacity
+	system capacity
+}
+
+// Capacity is node resource capacity
+type capacity struct {
+	hru uint64
+	sru uint64
+	cru uint64
+	mru uint64
+}
+
+// IsEmpty checks empty capacity
+func (cap *capacity) isEmpty() bool {
+	return cap.cru == 0 && cap.mru == 0 && cap.sru == 0 && cap.hru == 0
+}
+
+func (cap *capacity) update(c gridtypes.Capacity) {
+	cap.cru = c.CRU
+	cap.mru = uint64(c.MRU)
+	cap.sru = uint64(c.SRU)
+	cap.hru = uint64(c.HRU)
+}
+
+// Add adds a new for capacity
+func (cap *capacity) add(c capacity) {
+	cap.cru += c.cru
+	cap.mru += c.mru
+	cap.sru += c.sru
+	cap.hru += c.hru
+}
+
+// Subtract subtracts a new capacity
+func (cap *capacity) subtract(c capacity) (result capacity) {
+	result.cru = subtract(cap.cru, c.cru)
+	result.mru = subtract(cap.mru, c.mru)
+	result.sru = subtract(cap.sru, c.sru)
+	result.hru = subtract(cap.hru, c.hru)
+	return result
+}
+
+func subtract(a uint64, b uint64) uint64 {
+	if a > b {
+		return a - b
+	}
+	return 0
+}
+
+// Power represents power configuration
+type power struct {
+	WakeUpThreshold     uint8      `yaml:"wake_up_threshold"`
+	PeriodicWakeUpStart wakeUpDate `yaml:"periodic_wake_up_start"`
+	PeriodicWakeUpLimit uint8      `yaml:"periodic_wake_up_limit"`
+	OverProvisionCPU    float32    `yaml:"overprovision_cpu,omitempty"`
+}
+
+// wakeUpDate is the date to wake up all nodes
+type wakeUpDate time.Time
+
+// UnmarshalText unmarshal the given TOML string into wakeUp date
+func (d *wakeUpDate) UnmarshalText(b []byte) error {
```

Comment:
the goal is to unquote the double quotes around a time string? `"3:04PM"` ? please rename it to `fromQuotedDateTime` and maybe we should return another value instead of the pointer semantics here